src/compile.rs

```diff
@@ -682,6 +706,131 @@ fn check_required_config<'a>(
     ret.into_result(())
 }
 
+// Get the types, functions, and policy rules for the system.
+pub fn get_reduced_infos(
+    policies: &[PolicyFile],
+    classlist: &ClassList,
+    system: &ValidatedSystem,
+    type_map: &TypeMap,
+    module_map: &ModuleMap,
+) -> Result<Vec<sexp::Sexp>, CascadeErrors> {
+    let ret = CascadeErrors::new();
+    let mut new_type_map = get_built_in_types_map()?;
+
+    // Get the reduced type infos
+    for module in &system.modules {
+        get_reduced_types(module, &mut new_type_map, type_map, module_map)?;
+    }
+
+    // Generate type aliases for the new reduced type map
+    let new_t_aliases = collect_aliases(new_type_map.iter());
+    new_type_map.set_aliases(new_t_aliases);
+
+    // Get the function infos
+    let mut new_func_map = get_funcs(policies, &new_type_map)?;
+
+    // Validate functions
+    let new_func_map_copy = new_func_map.clone(); // In order to read function info while mutating
+    validate_functions(
+        &mut new_func_map,
+        &new_type_map,
+        classlist,
+        &new_func_map_copy,
+    )?;
+
+    // Get the policy rules
+    let new_policy_rules = get_policy_rules(policies, &new_type_map, classlist, &new_func_map)?;
+
+    let new_cil_tree = generate_sexp(
+        &new_type_map,
+        classlist,
+        new_policy_rules,
+        &new_func_map,
+        &Some(&system.configurations),
+    )?;
+
+    ret.into_result(new_cil_tree)
+}
+
+// This is a recusive function that gets only the relevant types from the type map.
+// The reduced types are the types in the module and the types in any of that modules' child modules.
+// Parents of those types are also automatically included.
+// Types are cloned for ownership.
```

Comment:
By "we clone the types because ownership", I more meant something like "Makes a clone of the types so that each system TypeMap can own its own types"